stan/math/prim/mat/prob/bernoulli_logit_glm_lpmf.hpp

```diff
@@ -0,0 +1,181 @@
+        stan::length(alpha)))
+    return 0.0;
+
+  T_partials_return logp(0.0);
+
+  check_bounded(function, "n", n, 0, 1);
+  check_not_nan(function, "Data matrix", x);
+  check_not_nan(function, "Weight vector", beta);
+  check_not_nan(function, "Bias vector", alpha);
+  check_consistent_sizes(function,
+                         "Random variable", n,
+                         "Bias vector", alpha);
+  // check_consistent_sizes(function,
+  //                       "Random variable", n,
+  //                       "Data matrix", x);  // would this work?
+  //                        maybe add one more check to see that x is
+  //                        compatible with beta?
+
+  if (!include_summand<propto, T_x, T_beta, T_alpha>::value)
+    return 0.0;
+
+  scalar_seq_view<T_n> n_vec(n);
+  scalar_seq_view<T_beta> beta_vec(beta);
+  scalar_seq_view<T_alpha> alpha_vec(alpha);
+
+  const size_t N = max_size(n, alpha);
+  // do we also want to check
+  // anything about x here? and in the next line?
+  const size_t M = beta.size();
+
+  operands_and_partials<T_x, T_beta, T_alpha> ops_partials(x, beta, alpha);
+
+  const bool constant_x = is_constant_struct<T_x>::value;
+  const bool constant_beta = is_constant_struct<T_beta>::value;
+  const bool constant_alpha = is_constant_struct<T_alpha>::value;
+
+  Matrix<double, Dynamic, 1> beta_dbl(M, 1);
+  Matrix<double, Dynamic, Dynamic> x_dbl(N, M);
+  for (size_t m = 0; m < M; m++)
+  {
+    beta_dbl[m] = value_of(beta_vec[m]);
+  }
+
+  for (size_t n = 0; n < N; n++)
+  { // could we vectorise this loop?
+    for (size_t m = 0; m < M; m++)
+    {
+      x_dbl(n, m) = value_of(x(n, m));
+    }
+    const int n_int = value_of(n_vec[n]);
+    const T_partials_return theta_dbl = (x_dbl.row(n) * beta_dbl)[0] +
+                                        value_of(alpha_vec[n]);
+
+    const int sign = 2 * n_int - 1;
+    const T_partials_return ntheta = sign * theta_dbl;
+    const T_partials_return exp_m_ntheta = exp(-ntheta);
+
+    // Handle extreme values gracefully using Taylor approximations.
+    static const double cutoff = 20.0;
+    if (ntheta > cutoff)
+      logp -= exp_m_ntheta;
+    else if (ntheta < -cutoff)
+      logp += ntheta;
+    else
+      logp -= log1p(exp_m_ntheta);
+
+    if (!(constant_x && constant_beta && constant_alpha))
+    {
+      static const double cutoff = 20.0; // do we really need this line?
+      T_partials_return theta_derivative;
+      if (ntheta > cutoff)
+        theta_derivative = -exp_m_ntheta;
+      else if (ntheta < -cutoff)
+        theta_derivative = sign;
+      else
+        theta_derivative = sign * exp_m_ntheta / (exp_m_ntheta + 1);
+      if (!constant_beta)
+      {
+        ops_partials.edge2_.partials_.col(0) += theta_derivative *
+                                                x_dbl.row(n).transpose();
+      }
+      if (!constant_x)
+      {
+        ops_partials.edge1_.partials_.row(n).noalias() += theta_derivative *
```

Comment:
What is `noalias()` doing here?  Should we be using this in other places?  Is there a way to turn these into matrix operations?

Reply:
The `noalias()` should lead to faster evaluation of matrix products if there is no aliasing in the equation. I've added it in another spot as well where it could speed things up now. I doubt it would make a huge difference though.

I guess I could try to make this whole loop into a matrix operation. I don't see a way of doing that while still maintaining the same handling of extreme values as bernoulli_logit_lpmf has, though. It seems to me that the current code is probably reasonably fast already, but I can't estimate how much vectorising this loop would speed things up. One problem would be that the new primitives wouldn't compute exactly the same as the old ones anymore then, which would complicate testing.

What would you like me to do?

Reply:
I think we should opt for precision over speed.